src/coreclr/jit/llvm.cpp

```diff
@@ -2387,6 +2427,55 @@ void Llvm::lowerCallToShadowStack(GenTreeCall* callNode)
     }
 }
 
+void Llvm::convertLclStructToLoad(GenTreeLclVarCommon* lclNode, ClassLayout* clsLayout)
+{
+    GenTree* lclAddr = _compiler->gtNewLclVarAddrNode(lclNode->GetLclNum());
+
+    lclNode->ChangeOper(GT_OBJ);
+    lclNode->AsObj()->SetAddr(lclAddr);
+    lclNode->AsObj()->SetLayout(clsLayout);
+
+    _currentRange->InsertBefore(lclNode, lclAddr);
+}
+
+void Llvm::lowerStoreLcl(GenTreeLclVarCommon* storeLclNode)
+{
+    GenTree* dataOp = storeLclNode->gtGetOp1();
+    CORINFO_CLASS_HANDLE dataHandle = _compiler->gtGetStructHandleIfPresent(dataOp);
+    if (dataOp->OperIs(GT_IND))
+    {
+        // Special case: "gtGetStructHandleIfPresent" sometimes guesses the handle from
+        // field sequences, but we will always need to transform TYP_STRUCT INDs into OBJs.
+        dataHandle = NO_CLASS_HANDLE;
+    }
+
+    LclVarDsc* addrVarDsc = _compiler->lvaGetDesc(storeLclNode->GetLclNum());
+
+    if (storeLclNode->TypeIs(TYP_STRUCT) && addrVarDsc->GetStructHnd() != dataHandle)
+    {
+        if (dataOp->OperIsIndir())
+        {
+            dataOp->SetOper(GT_OBJ);
+            dataOp->AsObj()->SetLayout(addrVarDsc->GetLayout());
+        }
+        else if (dataOp->OperIs(GT_LCL_VAR)) // can get icon 0 here
+        {
+            GenTreeLclVarCommon* dataLcl = dataOp->AsLclVarCommon();
+            if (storeLclNode->TypeGet() == TYP_STRUCT && dataLcl->TypeGet() == TYP_STRUCT)
+            {
+                LclVarDsc* dataVarDsc = _compiler->lvaGetDesc(dataLcl->GetLclNum());
+
+                if (tryGetStructClassHandle(addrVarDsc) != tryGetStructClassHandle(dataVarDsc))
+                {
+                    dataVarDsc->lvHasLocalAddr = 1;
+
+                    convertLclStructToLoad(dataLcl, addrVarDsc->GetLayout());
```

Comment:
I think `convertLclStructToLoad` can be inlined into this method (can be outlined again if needed in the future).